frame/contracts/src/migration/v12.rs

```diff
@@ -0,0 +1,324 @@
+	const VERSION: u16 = 12;
+
+	fn max_step_weight() -> Weight {
+		T::WeightInfo::v12_migration_step(T::MaxCodeLen::get())
+	}
+
+	fn step(&mut self) -> (IsFinished, Weight) {
+		let mut iter = if let Some(last_key) = self.last_key.take() {
+			PristineCode::<T>::iter_from(last_key.to_vec())
+		} else {
+			PristineCode::<T>::iter()
+		};
+		if let Some((hash, code)) = iter.next() {
+			let old = old::OwnerInfoOf::<T>::take(hash)
+				.expect(format!("OwnerInfo for code_hash {:?} not found!", hash).as_str());
+
+			log::debug!(target: LOG_TARGET, "Migrating OwnerInfo for code_hash {:?}", hash);
+
+			let module = old::CodeStorage::<T>::take(hash)
+				.expect(format!("No PrefabWasmModule found for code_hash: {:?}", hash).as_str());
+
+			// We print this to measure the impact of the migration.
+			// Storage removed: deleted PrefabWasmModule's encoded len.
+			// Storage added: determinism field encoded len (as all other CodeInfo fields are the
+			// same as in the deleted OwnerInfo).
+			log::debug!(
+				target: LOG_TARGET,
+				"Storage removed: 1 item, {} bytes",
+				module.encoded_size().saturating_sub(Determinism::max_encoded_len())
+			);
+
+			// Storage usage prices could change over time, and accounts who uploaded their
+			// conctracts code before the storage deposits where introduced, had not been ever
+			// charged with any deposit for that (see migration v6).
+			//
+			// This is why deposit to be refunded here is calculated as follows:
+			//
+			// 1. Calculate the deposit amount for storage before the migration, given current
+			// prices.
+			// 2. Given current reserved deposit amount, calculate the correction factor.
+			// 3. Calculate the deposit amount for storage after the migration, given current
+			// prices.
+			// 4. Calculate real deposit amount to be reserved after the migration.
+			let price_per_byte = T::DepositPerByte::get();
+			let price_per_item = T::DepositPerItem::get();
+			let bytes_before = module
+				.encoded_size()
+				.saturating_add(code.len())
+				.saturating_add(old::OwnerInfo::<T>::max_encoded_len()) as u32;
+			let items_before = 3u32;
+			let deposit_expected_before = price_per_byte
+				.saturating_mul(bytes_before.into())
+				.saturating_add(price_per_item.saturating_mul(items_before.into()));
+			let ratio = FixedU128::checked_from_rational(old.deposit, deposit_expected_before)
+				.unwrap_or_default()
+				.min(FixedU128::from_u32(1));
+			let bytes_after = code.len().saturating_add(CodeInfo::<T>::max_encoded_len()) as u32;
+			let items_after = 2u32;
+			let deposit_expected_after = price_per_byte
+				.saturating_mul(bytes_after.into())
+				.saturating_add(price_per_item.saturating_mul(items_after.into()));
+			let deposit = ratio.saturating_mul_int(deposit_expected_after);
+
+			let info = CodeInfo::<T> {
+				determinism: module.determinism,
+				owner: old.owner,
+				deposit,
+				refcount: old.refcount,
+			};
+
+			let amount = old.deposit.saturating_sub(info.deposit);
```

Comment:
might be a better fit for checked_sub 
```rust
if Some(amount) = old.deposit.checked_sub(info.deposit) {
...
} else {
...
}
```

Reply:
hm.. but when `amount == 0` we don't want to unreserve, right?

Reply:
> you might not want to warn if it's 0 too

why not? if it's zero it means that no deposit has been refunded, which is unexpected because this migration should pay deposit **every** contract owner by definiton (we free storage for every stored contract)